lib/class-wp-webfonts.php

```diff
@@ -0,0 +1,309 @@
+				trigger_error( __( 'Webfont src must be a non-empty string or an array of strings.', 'gutenberg' ) );
+				return false;
+			}
+		}
+
+		// Validate the 'src' property.
+		if ( ! empty( $font['src'] ) ) {
+			foreach ( (array) $font['src'] as $src ) {
+				if ( empty( $src ) || ! is_string( $src ) ) {
+					trigger_error( __( 'Each webfont src must be a non-empty string.', 'gutenberg' ) );
+					return false;
+				}
+			}
+		}
+
+		// Check the font-weight.
+		if ( ! is_string( $font['font-weight'] ) && ! is_int( $font['font-weight'] ) ) {
+			trigger_error( __( 'Webfont font weight must be a properly formatted string or integer.', 'gutenberg' ) );
+			return false;
+		}
+
+		// Check the font-display.
+		if ( ! in_array( $font['font-display'], array( 'auto', 'block', 'fallback', 'swap' ), true ) ) {
+			$font['font-display'] = 'fallback';
+		}
+
+		$valid_props = array(
+			'ascend-override',
+			'descend-override',
+			'font-display',
+			'font-family',
+			'font-stretch',
+			'font-style',
+			'font-weight',
+			'font-variant',
+			'font-feature-settings',
+			'font-variation-settings',
+			'line-gap-override',
+			'size-adjust',
+			'src',
+			'unicode-range',
+
+			// Exceptions.
+			'provider',
+		);
+
+		foreach ( $font as $prop => $value ) {
+			if ( ! in_array( $prop, $valid_props, true ) ) {
+				unset( $font[ $prop ] );
+			}
+		}
+
+		return $font;
+	}
+
+	/**
+	 * Register a provider.
+	 *
+	 * @param string $provider The provider name.
+	 * @param string $class    The provider class name.
+	 *
+	 * @return bool Whether the provider was registered successfully.
+	 */
+	public function register_provider( $provider, $class ) {
+		if ( empty( $provider ) || empty( $class ) ) {
+			return false;
+		}
+		self::$providers[ $provider ] = $class;
+		return true;
+	}
+
+	/**
+	 * Generate and enqueue webfonts styles.
+	 */
+	public function generate_and_enqueue_styles() {
+		// Generate the styles.
+		$styles = $this->generate_styles();
+
+		// Bail out if there are no styles to enqueue.
+		if ( '' === $styles ) {
+			return;
+		}
+
+		// Enqueue the stylesheet.
+		wp_register_style( $this->stylesheet_handle, '' );
+		wp_enqueue_style( $this->stylesheet_handle );
+
+		// Add the styles to the stylesheet.
+		wp_add_inline_style( $this->stylesheet_handle, $styles );
+	}
+
+	/**
+	 * Generate and enqueue editor styles.
+	 */
+	public function generate_and_enqueue_editor_styles() {
+		// Generate the styles.
+		$styles = $this->generate_styles();
+
+		// Bail out if there are no styles to enqueue.
+		if ( '' === $styles ) {
+			return;
+		}
+
+		wp_add_inline_style( 'wp-block-library', $styles );
+	}
+
+	/**
+	 * Generate styles for webfonts.
+	 *
+	 * By default (due to privacy concerns), this API will not do remote requests to
+	 * external webfont services nor generate `@font-face` styles for these remote
+	 * providers. The filter `'has_remote_webfonts_request_permission'` is provided
```

Comment:
These inline docs might be out of date. Are we still providing the `has_remote_webfonts_request_permission` filter?

Reply:
Thank you for catching this! Pushed a commit and removed these 👍